client/blocks/payment-request/hooks.js

```diff
@@ -0,0 +1,392 @@
+/* global wc_stripe_payment_request_params */
+
+/**
+ * External dependencies
+ */
+import { useState, useEffect, useCallback } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+import $ from 'jquery';
+
+/**
+ * Internal dependencies
+ */
+import {
+	getCartDetails,
+	createPaymentRequest,
+	updateShippingOptions,
+	updateShippingDetails,
+	createOrder,
+} from '../../api';
+import {
+	updatePaymentRequest,
+	normalizeShippingAddressForCheckout,
+} from '../stripe-utils';
+
+/**
+ * This hook takes care of creating a payment request and making sure
+ * you can pay through said payment request.
+ *
+ * @param {Object} stripe The stripe object used to create the payment request.
+ *
+ * @return {Array} An array; first element is the payment request; second element is the payment
+ *                 requests type.
+ */
+export const usePaymentRequest = ( stripe ) => {
+	const [ paymentRequest, setPaymentRequest ] = useState( null );
+	const [ paymentRequestType, setPaymentRequestType ] = useState( null );
+	useEffect( () => {
+		// Do nothing if Stripe object isn't loaded or paymentRequest already exists.
+		if ( ! stripe || paymentRequest ) {
+			return;
+		}
+
+		getCartDetails().then( ( cart ) => {
+			const pr = createPaymentRequest( stripe, cart );
+
+			pr.canMakePayment().then( ( result ) => {
+				if ( result ) {
+					setPaymentRequest( pr );
+					setPaymentRequestType(
+						result.applePay ? 'apple_pay' : 'payment_request_api'
+					);
+				}
+			} );
+		} );
+	}, [ paymentRequest, stripe ] );
+	return [ paymentRequest, paymentRequestType ];
+};
+
+/**
+ * Adds a shipping address change event handler to the provided payment request. Updates the
+ * order's shipping address when necessary.
+ *
+ * @param {Object} paymentRequest - The payment request object.
+ * @param {string} paymentRequestType - The payment request type.
+ * @param {Function} setShippingAddress - Used to set the shippingaddress in the Block.
+ */
+export const useShippingAddressUpdateHandler = (
+	paymentRequest,
+	paymentRequestType,
+	setShippingAddress
+) => {
+	useEffect( () => {
+		// Need to use `?.` here in case paymentRequest is null.
+		const shippingAddressUpdateHandler = paymentRequest?.on(
+			'shippingaddresschange',
+			( evt ) => {
+				const { shippingAddress } = evt;
+
+				// Update the block shipping address, just in case the payment request
+				// is cancelled.
+				setShippingAddress(
+					normalizeShippingAddressForCheckout( shippingAddress )
+				);
```

Comment:
Is there a reason for this `setShippingAddress`?

![image](https://user-images.githubusercontent.com/5509901/117502131-754d1900-af55-11eb-8d7c-b2053950c339.png)

This ends up making a second request to the Gutenberg endpoint, which leads to an error in those edge cases where the ZIP code is redacted or the state is not what WooCommerce is expecting (e.g Chinese provinces).

IMO the address choosing inside the payment request flow can be separate from the cart block shipping calculation. I don't know how we could manage to do it otherwise, given these inconsistencies between the address validation in the Payment Request API and what WooCommerce expects.

Reply:
Yeah that's an excellent point and not something I considered!
I agree, it's better (at least for now) to leave these 2 separated.

Removed in f47a87bf9a53ece85fa8417ec625e444f289edbc